src/Components/Web/src/Virtualization/Virtualize.cs

```diff
@@ -0,0 +1,264 @@
+            foreach (var item in _loadedItems.Skip(_itemsBefore).Take(_itemsVisible))
+            {
+                builder.AddContent(0, _itemTemplate, item);
+                itemsRendered++;
+            }
+
+            // Render placholder items to fill the remaining space.
+            for (; itemsRendered < _itemsVisible; itemsRendered++)
+            {
+                builder.AddContent(0, _placeholder, _itemsBefore + itemsRendered);
+            }
+
+            builder.CloseRegion();
+
+            builder.OpenElement(4, "div");
+            builder.AddAttribute(5, "style", GetSpacerStyle(itemsAfter));
+            builder.AddElementReferenceCapture(6, elementReference => _spacerAfter = elementReference);
+
+            builder.CloseElement();
+        }
+
+        private string GetSpacerStyle(int itemsInSpacer)
+            => $"height: {itemsInSpacer * ItemSize}px;";
+
+        void IVirtualizeJsCallbacks.OnBeforeSpacerVisible(float spacerSize, float containerSize)
+        {
+            _itemsBefore = CalcualteItemDistribution(spacerSize, containerSize);
+
+            StateHasChanged();
+        }
+
+        void IVirtualizeJsCallbacks.OnBottomSpacerVisible(float spacerSize, float containerSize)
+        {
+            var itemsAfter = CalcualteItemDistribution(spacerSize, containerSize);
+
+            _itemsBefore = Math.Max(0, _itemCount - itemsAfter - _itemsVisible);
+
+            if (_itemsBefore + _visibleItemCapacity > _loadedItems.Count)
+            {
+                _fetchTask = FetchItems();
+            }
+
+            StateHasChanged();
+        }
+
+        private int CalcualteItemDistribution(float spacerSize, float containerSize)
+        {
+            _visibleItemCapacity = (int)Math.Ceiling(containerSize / ItemSize) + 2;
+            _itemsVisible = Math.Clamp(_itemCount, 0, _visibleItemCapacity);
+
+            return Math.Max(0, (int)Math.Floor(spacerSize / ItemSize) - 1);
+        }
+
+        private async Task FetchItems()
+        {
+            // Cancel the previous fetch, if it exists.
+            _fetchCts?.Cancel();
+
+            // Wait for the task to complete. If it fails due to an exception, the exception will bubble-up
+            // through `OnAfterRenderAsync`.
+            if (!_fetchTask.IsCanceled)
+            {
+                await _fetchTask;
+            }
+
+            var placeholderItems = _itemsBefore + _visibleItemCapacity - _loadedItems.Count;
+
+            // There's a chance that no more items need to be fetched after the previous task completes.
+            if (placeholderItems <= 0)
+            {
+                return;
+            }
+
+            _fetchCts = new CancellationTokenSource();
+
+            var result = await _itemsProvider(new ItemsProviderRequest(_loadedItems.Count, placeholderItems, _fetchCts.Token));
+
+            _itemCount = result.TotalItemCount;
+            _loadedItems.AddRange(result.Items);
```

Comment:
I think we'll have an issue here, because this is going to accumulate items forever. If the user has (say) a list with 100000 items and scrolls all the way through it, we'd build up all 100000 in memory, plus it looks like we'd keep all 100000 element subtrees in the DOM.

You could choose to handle this by removing things from `_loadedItems` once we're sufficiently past them (though not in the infinite scroll case). Alternatively you could try to avoid having a `_loadedItems` collection at all, and only ever just keep hold of the most recent batch we received (along with its index offset into the source collection).

Reply:
Actually I'm super confused about this and am investigating to try to understand better! I don't quite get how it's legit to always append to the list like this, given we might be loading the items out of order (e.g., if a user scrolls instantly to the end of the list, then slowly starts scrolling backwards). Maybe there's more going on than I've understood!

Reply:
Ah right I see. From a bit of debugging I see it potentially loads massive chunks in order to keep the loaded set contiguous. For example I have declared a set with 100000 items and it just asked the provider for a chunk of length 5236 when I scrolled quickly.

We'll need to make it fetch only chunks needed to cover the visible range, which means the loaded set can't be assumed to be contiguous. More generally it reinforces the idea that we perhaps shouldn't have a "loaded set" other than the most recently-loaded chunk. The need for InfiniteScroll to continually add to a single collection is a behavior we'll need to limit to InfiniteScroll.

I appreciate this complicates things a bit. Sorry if that aspect of the requirements wasn't clear before. I'm having a look to see if I can make any useful suggestions about how this can be implemented without too much trouble.